lib/frame.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+class Frame
+  TEN_PINS = 10
+  FIRST_ROLL = 0
+  SECOND_ROLL = 1
+
+  def initialize
+    @rolls = []
+    @frame_type = nil
+  end
+
+  def log_roll(pins_downed)
+    raise 'Pins downed must be between 0 and 10' if invalid_pins?(pins_downed)
+    raise 'Frame complete. Cannot roll again' if frame_complete?
+
+    process_frame_type(pins_downed)
+    process_roll(pins_downed)
+  end
+
+  def sum_frame
+    @rolls.sum
+  end
+
+  def all_rolls
+    @rolls
+  end
+
+  def frame_complete?
+    @rolls.count > 1 || @frame_type == :strike
+  end
+
+  def strike_frame?
+    @frame_type == :strike
+  end
+
+  def spare_frame?
+    @frame_type == :spare
+  end
+
+  def open_frame?
+    @frame_type == :open
+  end
```

Comment:
Ideally we want to minimise the amount of state stored in a class. Whether the frame is a strike or spare can be calculated directly from other state (the rolls), so it's best not to duplicate it in a `@frame_type` variable.

Reply:
Instead, consider a strike? method which works out from the rolls whether or not the frame is a strike.

Reply:
Implemented in the javascript version. Thanks for the feedback